<html>
<META http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<head>
<title>Section 10.12.&nbsp; Exceptions and the sys Module</title>
<link rel="STYLESHEET" type="text/css" href="images/style.css">
<link rel="STYLESHEET" type="text/css" href="images/docsafari.css">
</head>
<body>
<table width="100%" border="0" cellspacing="0" cellpadding="0">
<tr><td><div STYLE="MARGIN-LEFT: 0.15in;">
<a href=ch10lev1sec11.html><img src="images/prev.gif" width="60" height="17" border="0" align="absmiddle" alt="Previous Page"></a>
<td align="right"><div STYLE="MARGIN-LEFT: 0.15in;">
<a href=ch10lev1sec13.html><img src="images/next.gif" width="60" height="17" border="0" align="absmiddle" alt="Next Page"></a>
</div></td></tr></table>
<br><table width="100%" border="0" cellspacing="0" cellpadding="0"><tr><TD valign="top"><a name="ch10lev1sec12"></a>
<h3 id="title-IDA4PBPB" class="docSection1Title">10.12. Exceptions and the <tt>sys</tt> Module</H3>
<p class="docText">An alternative way of obtaining exception information is by accessing the <tt>exc_info()</tt> function in the <tt>sys</tt> module. This function provides a 3-tuple of information, more than what we can achieve by simply using only the exception argument. Let us see what we get using <tt>sys.exc_info()</tt>:</p>
<div class="docText"><pre>&gt;&gt;&gt; <span class="docEmphStrong">try</span>:
...     float('abc123')
... <span class="docEmphStrong">except</span>:
...     <span class="docEmphStrong">import</span> sys
...        exc_tuple = sys.exc_info()
...
&gt;&gt;&gt; <span class="docEmphStrong">print</span> exc_tuple
(&lt;class exceptions.ValueError at f9838&gt;, &lt;exceptions.
ValueError instance at 122fa8&gt;,
&lt;traceback object at 10de18&gt;)
&gt;&gt;&gt;

&gt;&gt;&gt; <span class="docEmphStrong">for</span> eachItem <span class="docEmphStrong">in</span> exc_tuple:
...     <span class="docEmphStrong">print</span> eachItem
...
exceptions.ValueError
invalid literal for float(): abc123
&lt;traceback object at 10de18&gt;</pre></div><BR>
<p class="docText"><a name="iddle1866"></a>What we get from <tt>sys.exc_info()</tt> in a tuple are:</P>
<UL><LI><p class="docList"><tt>exc_type</tt>: exception class object</p></li><li><p class="docList"><tt>exc_value</tt>: (this) exception class instance object</P></li><LI><p class="docList"><tt>exc_traceback</tt>: traceback object</P></li></UL>
<p class="docText">The first two items we are familiar with: the actual exception class and this particular exception's instance (which is the same as the exception argument which we discussed in the previous section). The third item, a traceback object, is new. This object provides the execution context of where the exception occurred. It contains information such as the execution frame of the code that was running and the line number where the exception occurred.</P>
<p class="docText">In older versions of Python, these three values were available in the <tt>sys</tt> module as <tt>sys.exc_type</tt>, <tt>sys.exc_value</tt>, and <tt>sys.exc_traceback</tt>. Unfortunately, these three are global variables and not thread-safe. We recommend using <tt>sys.exc_info()</tt> instead. All three will be phased out and eventually removed in a future version of Python.</p>

</TD></TR></table>
<br>
<table width="100%" border="0" cellspacing="0" cellpadding="0">
<tr><td><div STYLE="MARGIN-LEFT: 0.15in;">
<a href=ch10lev1sec11.html><img src="images/prev.gif" width="60" height="17" border="0" align="absmiddle" alt="Previous Page"></a>
<td align="right"><div STYLE="MARGIN-LEFT: 0.15in;">
<a href=ch10lev1sec13.html><img src="images/next.gif" width="60" height="17" border="0" align="absmiddle" alt="Next Page"></a>
</div></td></tr></table>
</body></html>